octodns_constellix/__init__.py

```diff
@@ -20,78 +20,114 @@
-    def __init__(self, resp):
-        errors = '\n  - '.join(resp.json()['errors'])
+class ConstellixAPIBadRequest(ConstellixAPIException):
+    def __init__(self, data):
+        errors = '\n  - '.join(data.get('errors', []))
         super().__init__(f'\n  - {errors}')
 
 
-class ConstellixClientUnauthorized(ConstellixClientException):
+class ConstellixAPIUnauthorized(ConstellixAPIException):
     def __init__(self):
         super().__init__('Unauthorized')
 
 
-class ConstellixClientNotFound(ConstellixClientException):
+class ConstellixAPINotFound(ConstellixAPIException):
     def __init__(self):
         super().__init__('Not Found')
 
 
-class ConstellixClient(object):
-    BASE = 'https://api.dns.constellix.com/v1'
-
-    def __init__(self, log, api_key, secret_key, ratelimit_delay=0.0):
+class ConstellixAPI(object):
+    def __init__(self, base_url, log, api_key, secret_key, ratelimit_delay):
+        self.base_url = base_url
         self.log = log
         self.api_key = api_key
         self.secret_key = secret_key
         self.ratelimit_delay = ratelimit_delay
         self._sess = Session()
         self._sess.headers.update(
             {
-                'x-cnsdns-apiKey': self.api_key,
-                'User-Agent': f'octodns/{octodns_version} octodns-constellix/{__VERSION__}',
+                'User-Agent': f'octodns/{octodns_version} octodns-constellix/{__VERSION__}'
             }
         )
-        self._domains = None
-        self._pools = {'A': None, 'AAAA': None, 'CNAME': None}
-        self._geofilters = None
 
     def _current_time(self):
         return str(int(time.time() * 1000))
 
-    def _hmac_hash(self, now):
-        return hmac.new(
-            self.secret_key.encode('utf-8'),
-            now.encode('utf-8'),
-            digestmod=hashlib.sha1,
-        ).digest()
+    def _hmac_text(self, now):
+        return b64encode(
+            hmac.new(
+                self.secret_key.encode('utf-8'),
+                now.encode('utf-8'),
+                digestmod=hashlib.sha1,
+            ).digest()
+        )
+
+    def _url(self, path):
+        return f'{self.base_url}{path}'
+
+    def _get_json(self, response):
+        try:
+            return response.json()
+        except:
+            raise ConstellixAPIBadRequest({'errors': [response.text]})
 
     def _request(self, method, path, params=None, data=None):
-        self.log.debug("Call _request %s %s", method, path)
+        url = self._url(path)
+        self.log.debug('Call _request %s %s', method, url)
         now = self._current_time()
-        hmac_hash = self._hmac_hash(now)
+        hmac_text = self._hmac_text(now)
+        auth_token = f'{self.api_key}:{hmac_text}:{now}'
 
         headers = {
-            'x-cnsdns-hmac': b64encode(hmac_hash),
-            'x-cnsdns-requestDate': now,
+            'x-cns-security-token': auth_token,  # v1, v2, sonar
+            'authorization': f'Bearer {auth_token}',  # v4
```

Comment:
Interesting. Assume it doesn't hurt to define both/the server side will pick which it needs/prefers?

Reply:
It doesn't hurt. But of course it would be a clean solution to only provide one auth header. Will add this.

Reply:
Addressed in 28e7807845258801bef9483ab13d4d272a6e9d80